01_git/scissors/game.cpp

```diff
@@ -0,0 +1,149 @@
+#include <cstdlib>
+#include <iostream>
+
+// Not using enum class since too lazy to
+// do explicit casts to size_t everywhere.
+enum Choice
+{
+    SCISSORS = 0,
+    PAPER,
+    ROCK,
+    CHOICE_LAST
+};
+
+enum GameResult
+{
+    PLAYER_WINS = 0,
+    CPU_WINS,
+    DRAW,
+    GR_LAST
+};
+
+namespace
+{
+
+const char* gaItemsStrRepr[]
+{
+    "scissors", // SCISSORS
+    "paper",    // PAPER
+    "rock"      // ROCK
+};
+
+const char* gaGameResStrRepr[]
+{
+    "You win", // PLAYER_WINS
+    "I win",   // CPU_WINS
+    "Draw!"    // DRAW
+};
+
+GameResult gaResolveTbl[][CHOICE_LAST] {
+    //   SCISSORS    |     PAPER     |     ROCK      <- Player's choice
+    //
+    //                                                  CPU's choice
+    //                                                     |
+    //                                                     V
+    {    DRAW,          PLAYER_WINS,     CPU_WINS    },  // SCISSORS
+    {  CPU_WINS,           DRAW,        PLAYER_WINS  },  // PAPER
+    {  PLAYER_WINS,      CPU_WINS,         DRAW      }   // ROCK
+};
+
+void resolveGame(Choice rePlayerChoice, Choice reCpuChoice)
+{
+    const auto eResult = gaResolveTbl[rePlayerChoice][reCpuChoice];
+    std::cout << gaGameResStrRepr[eResult];
+
+    if (eResult != DRAW)
+    {
+        const auto eFirstNoun = (eResult == CPU_WINS) ? reCpuChoice : rePlayerChoice;
+        const auto eSecondNoun = (eFirstNoun == reCpuChoice) ? rePlayerChoice : reCpuChoice;
+        std::cout << ": " << gaItemsStrRepr[eFirstNoun];
+        std::cout << " beat" << (eFirstNoun != SCISSORS ? "s" : "");
+        std::cout << " " << gaItemsStrRepr[eSecondNoun];
+    }
+
+    std::cout << std::endl;
+}
+
+Choice decodeChoice(int rdRawChoice)
+{
+    switch (rdRawChoice)
+    {
+    case 's':
+        return SCISSORS;
+    case 'p':
+        return PAPER;
+    case 'r':
+        return ROCK;
+    default:
+        return CHOICE_LAST;
+    }
+}
+
+Choice makeCpuChoice()
+{
+    while (true)
+    {
+        const auto dRand = rand() % CHOICE_LAST;
+        if (dRand < CHOICE_LAST)
+        {
+            return Choice(dRand);
+        }
+    }
+}
+} // namespace
+
+int main()
+{
+    std::srand(time(nullptr));
+
+    int cInput = 0;
+    bool bSkipHeading = false;
+
+    while (true)
+    {
+        if (not bSkipHeading)
+        {
+            std::cout << "Please choose: rock (r) - paper (p) - scissors (s)" << std::endl;
+        }
+        else
+        {
+            bSkipHeading = false;
+        }
+
+        cInput = std::tolower(std::getchar());
```

Comment:
It will be better to read directly to the variable

std::cin >> user_option;

Reply:
Agree. Originally I had wanted to implement keycodes scanning later I decided to keep things simple. Will address this asap.